sdk/eventgrid/Azure.Messaging.EventGrid/src/Customization/CloudEvent.cs

```diff
@@ -0,0 +1,79 @@
+    public partial class CloudEvent
+    {
+        /// <summary> Initializes a new instance of CloudEvent. </summary>
+        /// <param name="source"> Identifies the context in which an event happened. The combination of id and source must be unique for each distinct event. </param>
+        /// <param name="type"> Type of event related to the originating occurrence. </param>
+        public CloudEvent(string source, string type)
+        {
+            if (source == null)
+            {
+                throw new ArgumentNullException(nameof(source));
+            }
+            if (type == null)
+            {
+                throw new ArgumentNullException(nameof(type));
+            }
+
+            Source = source;
+            Type = type;
+            ExtensionAttributes = new Dictionary<string, object>();
+        }
+
+        /// <summary> An identifier for the event. The combination of id and source must be unique for each distinct event. </summary>
+        public string Id { get; set; } = Guid.NewGuid().ToString();
+        /// <summary> Identifies the context in which an event happened. The combination of id and source must be unique for each distinct event. </summary>
+        public string Source { get; set; }
+        /// <summary> Event data specific to the event type. </summary>
+        public object Data { get; set; }
+        /// <summary> Type of event related to the originating occurrence. </summary>
+        public string Type { get; set; }
+        /// <summary> The time (in UTC) the event was generated, in RFC3339 format. </summary>
+        public DateTimeOffset? Time { get; set; } = DateTimeOffset.UtcNow;
+        /// <summary> The version of the CloudEvents specification which the event uses. </summary>
+        public string SpecVersion { get; set; } = "1.0";
+        /// <summary> Identifies the schema that data adheres to. </summary>
+        public string DataSchema { get; set; }
+        /// <summary> Content type of data value. </summary>
+        public string DataContentType { get; set; }
+        /// <summary> This describes the subject of the event in the context of the event producer (identified by source). </summary>
+        public string Subject { get; set; }
+        /// <summary>
+        /// Extension attributes that can be additionally added to the CloudEvent envelope.
+        /// </summary>
+        public Dictionary<string, object> ExtensionAttributes { get; }
+        /// <summary>
+        /// Adds or overwrites an extension attribute to the event envelope. Attribute must not have the same name as standard CloudEvent
+        /// properties (like 'id' and 'source'), and must follow the same naming convention (lower-case letters and/or numbers).
+        /// </summary>
+        /// <param name="key">Attribute name.</param>
+        /// <param name="value">Attribute value.</param>
+        public void AddExtensionAttribute(string key, object value)
```

Comment:
Hmm, why do we have these methods rather just relying on the user to directly add/remove from the property?

Reply:
Yeah I wasn't sure if we needed to have these methods. For the add method, I wanted the user to be able to overwrite a previously-added value and not have a thrown exception (since each key-value pair is like another property on the envelope, I though that the add method could act like a setter?). Also, I thought the summary for the add method explaining some of the naming restrictions could be helpful for the user. But I agree the remove method isn't really necessary.

Reply:
I think if we were to actually provide validation, then it could make sense to have these methods. But I don't think we want to do that as these are service level constraints. In general, when we have a dictionary property on a model, we initialize it and provide a getter. The user is responsible for adding and removing any elements.

Reply:
Ok that makes sense!